sphinx_gallery/gen_gallery.py

```diff
@@ -687,6 +691,199 @@ def write_computation_times(gallery_conf, target_dir, costs):
             fid.write(hline)
 
 
+def write_api_entries(app, what, name, obj, options, lines):
+    if 'api_entries' not in app.config.sphinx_gallery_conf:
+        app.config.sphinx_gallery_conf['api_entries'] = \
+            {entry_type: set() for entry_type in
+             ('class', 'method', 'function', 'module',
+              'property', 'attribute')}
+    app.config.sphinx_gallery_conf['api_entries'][what].add(name)
+
+
+def init_api_usage(gallery_dir):
+    with codecs.open(os.path.join(gallery_dir, 'sg_api_usage.rst'), 'w',
+                     encoding='utf-8'):
+        pass
+
+
+def write_api_entry_usage(app, docname, source):
+    gallery_conf = app.config.sphinx_gallery_conf
+    # since this is done at the gallery directory level (as opposed
+    # to in a gallery directory, e.g. auto_examples), it runs last
+    # which means that all the api entries will be in gallery_conf
+    if 'sg_api_usage' not in docname or \
+            'api_entries' not in gallery_conf or \
+            gallery_conf['backreferences_dir'] is None:
+        return
+    backreferences_dir = os.path.join(gallery_conf['src_dir'],
+                                      gallery_conf['backreferences_dir'])
+    try:
+        import graphviz
+        has_graphviz = True
+    except ImportError:
+        logger.info('`graphviz` required for graphical visualization')
+        has_graphviz = False
+
+    example_files = set.union(
+        *[gallery_conf['api_entries'][obj_type]
+          for obj_type in ('class', 'method', 'function')
+          if obj_type in gallery_conf['api_entries']])
+
+    total_count = len(example_files)
+    if total_count == 0:
+        return
+
+    def get_entry_type(entry):
+        if entry in gallery_conf['api_entries']['class']:
+            return 'class'
+        elif entry in gallery_conf['api_entries']['method']:
+            return 'meth'
+        else:
+            assert entry in gallery_conf['api_entries']['function']
+            return 'func'
+
+    # find used and unused API entries
+    unused_api_entries = list()
+    used_api_entries = dict()
+    for entry in example_files:
+        if '__' in entry:  # don't include built-in methods
```

Comment:
Even in MNE-Python we might want to know if things like `__call__` are used. It might be better to expose this as a regex for entries to ignore like `missing_doc_ignore='__.*__'`.

This can also be a way to allow people to disable this -- if people use `'.*'` then all entries will match, and you can special case that `if len(used_api_entries) == 0` then don't write anything.

Reply:
Could we save this for a follow-up?

Reply:
I think we want to make it optional *somehow* in this PR before merge. And however it works needs to be documented. So actually the `re.match(pattern, entry) is not None` change on this line is not much to add on top of those requirements

Reply:
Ok, done

Reply:
> Ok, done

Not quite, reading a bit more closely above you'll see

>> And however it works needs to be documented.

i.e., you need to add to `doc/configuration.rst`

Reply:
Ok, sorry I missed that.